File: online-training-platform/leetcode/290. Word Pattern.cpp
class Solution {
public:
    
    vector<string> split(const string& s, char delimiter){
        vector<string> tokens;
        string token;
        istringstream tokenStream(s);
        while (getline(tokenStream, token, delimiter)){
            tokens.push_back(token);
        }
        return tokens;
    }
    
    bool wordPattern(string pattern, string str) {
        map<char, string> m;
        vector<string> words = split(str, ' ');
        if(pattern.length() != words.size()) return false;
        int size = words.size();
        string word;
        char character;
        for(int i = 0 ; i < size; i++){
            word = words[i];
            character = pattern[i];
            if(m.count(character) > 0){
                if(m[character] != word) return false;
            }else{
                for(map<char,string>::iterator it = m.begin() ; it != m.end() ; it++){
                    if(it->second == word) return false;
                }
        
                m[character] = word;  
            }
        }
        
        
        return true;
    }
};

File: online-training-platform/leetcode/290. Word Pattern.kt
class Solution {
    fun wordPattern(pattern: String, s: String): Boolean {
        val words = s.split(" ")
        val wordToLetter = HashMap<String, Char>()
        val letterToWord = HashMap<Char, String>()
        
        if(words.size != pattern.length) {
            return false
        }
        
        val size = words.size
        words.forEachIndexed { idx, word ->
            if(wordToLetter.contains(word)){
                if(wordToLetter.getOrDefault(word, " ") != pattern[idx]){
                    return false
                }
            }
            
            if(letterToWord.contains(pattern[idx])){
                if(letterToWord.getOrDefault(pattern[idx], ' ') != word) {
                    return false
                }
            }
            
            wordToLetter[word] = pattern[idx]
            letterToWord[pattern[idx]] = word
        }
        
        return true
    }
}